# SCAPP    
SCAPP is a subproject of PCmaster.    
SCAPP == PCmaster_ann.    
    
PCmaster    
Plant Cell master    

PCmaster_ann    
Plant Cell master for automatic annotation of cell types    

SCAPP    
Single-Cell Aid Project for Plants on Python       
Single-Cell Auto Annotation Project for Plants on Python

Version 1.0.0      
You can download the required files on https://pan.baidu.com/s/1qwpAcdd4NALRgTtTtMnnuw?pwd=wkwr.    
   
If you want to use the latest version, please download the latest 'SCAPP_XX_XX_XX.py' file and copy the code in it to replace the code in 'SCAPP_0_copy1.py'.     
Please note that the latest version of the code may cause some undetected bugs.

Sincerely thanks to the contributors of python packages such as d2l, doubletdetection, harmonypy, numpy, pandas, scanpy, seaborn, scrublet, scikit-learn, torch, etc.

If you have questions about scapp, please send email to 12216017@zju.edu.cn.

Main References:    
Brown, T.B., Mann, B., Ryder, N., Subbiah, M., Kaplan, J., Dhariwal, P., Neelakantan, A., Shyam, P., Sastry, G., Askell, A., Agarwal, S., Herbert-Voss, A., Krueger, G., Henighan, T., Child, R., Ramesh, A., Ziegler, D.M., Wu, J., Winter, C., Hesse, C., Chen, M., Sigler, E., Litwin, M., Gray, S., Chess, B., Clark, J., Berner, C., McCandlish, S., Radford, A., Sutskever, I., Amodei, D., (2020). Language Models are Few-Shot Learners. ArXiv.    
Harris, C.R., Millman, K.J., van der Walt, S.J., Gommers, R., Virtanen, P., Cournapeau, D., Wieser, E., Taylor, J., Berg, S., Smith, N.J., Kern, R., Picus, M., Hoyer, S., van Kerkwijk, M.H., Brett, M., Haldane, A., del Río, J.F., Wiebe, M., Peterson, P., Gérard-Marchant, P., Sheppard, K., Reddy, T., Weckesser, W., Abbasi, H., Gohlke, C., Oliphant, T.E., (2020). Array programming with NumPy. Nature 585, 357–362. https://doi.org/10.1038/s41586-020-2649-2    
Korsunsky, I., Millard, N., Fan, J., Slowikowski, K., Zhang, F., Wei, K., Baglaenko, Y., Brenner, M., Loh, P., Raychaudhuri, S., (2019). Fast, sensitive, and accurate integration of single cell data with Harmony. Nat Methods 16, 1289–1296. https://doi.org/10.1038/s41592-019-0619-0    
Ortiz-Ramírez, C., Guillotin, B., Xu, X., Rahni, R., Zhang, S., Yan, Z., Coqueiro Dias Araujo, P., Demesa-Arevalo, E., Lee, L., Van Eck, J., Gingeras, T.R., Jackson, D., Gallagher, K.L., Birnbaum, K.D., (2021). Ground tissue circuitry regulates organ complexity in maize and Setaria. Science 374, 1247–1252. https://doi.org/10.1126/science.abj2327    
Ouyang, L., Wu, J., Jiang, X., Almeida, D., Wainwright, C.L., Mishkin, P., Zhang, C., Agarwal, S., Slama, K., Ray, A., Schulman, J., Hilton, J., Kelton, F., Miller, L., Simens, M., Askell, A., Welinder, P., Christiano, P., Leike, J., Lowe, R., (2022). Training language models to follow instructions with human feedback. https://doi.org/10.48550/arXiv.2203.02155    
Shahan, R., Hsu, C.-W., Nolan, T.M., Cole, B.J., Taylor, I.W., Greenstreet, L., Zhang, S., Afanassiev, A., Vlot, A.H.C., Schiebinger, G., Benfey, P.N., Ohler, U., (2022). A single-cell Arabidopsis root atlas reveals developmental trajectories in wild-type and cell identity mutants. Dev Cell 57, 543-560.e9. https://doi.org/10.1016/j.devcel.2022.01.008    
Wang, Y., Huan, Q., Li, K., Qian, W., (2021). Single-cell transcriptome atlas of the leaf and root of rice seedlings. J Genet Genomics 48, 881–898. https://doi.org/10.1016/j.jgg.2021.06.001    
Wolf, F.A., Angerer, P., Theis, F.J., (2018). SCANPY: large-scale single-cell gene expression data analysis. Genome Biol 19, 15. https://doi.org/10.1186/s13059-017-1382-0    
Wolock, S.L., Lopez, R., Klein, A.M., (2019). Scrublet: Computational Identification of Cell Doublets in Single-Cell Transcriptomic Data. Cell Syst 8, 281-291.e9. https://doi.org/10.1016/j.cels.2018.11.005    
Xi, N.M., Li, J.J., (2021). Protocol for executing and benchmarking eight computational doublet-detection methods in single-cell RNA sequencing data analysis. STAR Protoc 2, 100699. https://doi.org/10.1016/j.xpro.2021.100699    
Zong, J., Wang, L., Zhu, L., Bian, L., Zhang, B., Chen, X., Huang, G., Zhang, X., Fan, J., Cao, L., Coupland, G., Liang, W., Zhang, D., Yuan, Z., (2022). A rice single cell transcriptomic atlas defines the developmental trajectories of rice floret and inflorescence meristems. New Phytologist 234, 494–512. https://doi.org/10.1111/nph.18008    
   
    
D2l. https://zh-v2.d2l.ai/    
Pandas. https://pandas.pydata.org/docs/index.html    
PyTorch. https://pytorch.org/    
Scikit-learn. https://scikit-learn.org/stable/    
